**Иван, привет!**

Меня зовут Алексей Каргин, я буду проверять проект по **SQL** в рамках финального спринта. Приятно познакомиться! Предлагаю общаться на «ты». Если это неприемлемо, будет здорово, если я об этом узнаю, и мы перейдем на «вы». Для удобства я оставлю комментарии в отдельных ячейках Markdown с заголовком «Комментарий ревьюера». Пожалуйста, не перемещай, не изменяй и не удаляй их - так наше общение будет более продуктивным. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера</b> 
    
✍ Так выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера</b> 
    
📝 Так выделены небольшие замечания или рекомендации. Постарайся учесть эти комментарии в этом проекте.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера</b> 
    
👍 Так выделены все остальные комментарии, включая рекомендации, позитивные моменты или какие-то рассуждения.
</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:

<div class="alert alert-info"> 
<b>Комментарий студента</b> 
    
Например, вот так.
</div>

---

<div class="alert alert-success">
<b>Обратная связь v.1</b> 
    
👋 Иван, спасибо, что прислал проект! Ты хорошо придерживаешься оформления проекта, твои запросы легко читать, и они также хорошо оформлены. Мы можем также добавлять небольшие комментарии к подзапросам - так проще возвращаться к своим запросам и проектам через некоторое время.
    
<br>
<div class="text-danger">   
    
✍ Однако проект пока нельзя принять, так как есть несколько моментов во втором запросе, которые нужно проверить - возможно сейчас мы дублируем данные по количеству обзоров и мы выводим информацию не для всех книг (если брать по id).

    
    
</div>    

    
Посмотри, пожалуйста, мои дополнительные комментарии - там больше деталей, плюс есть иные рекомендации. Мне кажется, что несмотря на все мои комментарии у тебя не составит труда внести необходимые изменения. Буду ждать новой версии проекта!

<br>    
С наилучшими пожеланиями, <br>
Алексей
</div>


---

<div class="alert alert-success">
<b>Обратная связь v.2</b> 
    
👋 Иван, спасибо, спасибо, что оперативно внес корректировку в запросы - в целом, теперь все верно и проект может быть принят. Однако, у тебя были некоторые вопросы, по ходу проекта, посмотри, пожалуйста, мои комментарии - если у тебя больше нет вопросов, то дай знать, чтобы я принял проект.
    
<br>    
С наилучшими пожеланиями, <br>
Алексей
</div>

---

<div class="alert alert-success">
<b>Обратная связь v.3</b> 
    
🎉 Иван, теперь, когда вопросов нет, могу только поздравить с окончанием финального спринта и всего обучения на курсе Яндекс Практикума! Спасибо за работу над проектом - у тебя отличная мотивация изучать новое и совершенствовать свои навыки!
    
    
<br>    
С наилучшими пожеланиями, <br>
Алексей
</div>

---

# Первичный анализ рынка книг

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![https://s3-us-west-2.amazonaws.com/secure.notion-static.com/069818d1-0e5c-4d87-a461-0de584ab9c33/Untitled_(33).png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/069818d1-0e5c-4d87-a461-0de584ab9c33/Untitled_(33).png)

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 Хорошо, что есть краткое описание проекта и исходных таблиц, а также перечислены решаемые задачи.
</div>

## Импорт библиотек

In [1]:
#импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

## Подключение к базе данных

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b>
    
👍 Все готово для подключения к базе данных! 

</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера v.1</b> 
    
📝 Лучше отделить импорт библиотек от остального кода и вынести его в отдельную ячейку. Это соответствует стандарту оформления кода [PEP8](https://www.python.org/dev/peps/pep-0008/#imports).
    
</div>

<div class="alert alert-info">
<b>Комментарий студента v.1</b>
    
Вынес

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 Спасибо!
</div>

## Выполнение запросов

Для начала изучим каждую из таблиц

In [3]:
queryBooks = '''
SELECT
    *
FROM
    books
'''
print('Таблица BOOKS')
pd.io.sql.read_sql(queryBooks, con = engine).head()

Таблица BOOKS


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
queryAuthors = '''
SELECT
    *
FROM
    authors
'''
print('Таблица AUTHORS')
pd.io.sql.read_sql(queryAuthors, con = engine).head()

Таблица AUTHORS


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
queryRatings = '''
SELECT
    *
FROM
    ratings
'''
print('Таблица RATINGS')
pd.io.sql.read_sql(queryRatings, con = engine).head()

Таблица RATINGS


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
queryReviews = '''
SELECT
    *
FROM
    reviews
'''
print('Таблица REVIEWS')
pd.io.sql.read_sql(queryReviews, con = engine).head()

Таблица REVIEWS


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [7]:
queryPublishers = '''
SELECT
    *
FROM
    publishers
'''
print('Таблица PUBLISHERS')
pd.io.sql.read_sql(queryPublishers, con = engine).head()

Таблица PUBLISHERS


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Самая интересная таблица - *books* - она связывает оставшиеся четыре таблицы между собой. С ней и начнем работу

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b>
    
👍 Отлично! Подключение к базе  данных работает, таблицы на месте. На экран были выведены первые строки из каждой таблицы, что позволило оценить их содержание. Однако, обрати, пожалуйста, внимание, что в запросе мы загружаем все данные, поэтому перед их выгрузкой лучше оценить их объем. Почему это важно - если данных много, а их может быть много и они могут быть объемными, мы можем создать дополнительную нагрузку на окружение, что может быть не очень хорошо.
</div>

<div class="alert alert-info">
<b>Комментарий студента v.1</b>
    
А как оценивать объем? COUNT * ?

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 Да, это самый простой способ. 
</div>

### Сколько книг вышло после 1 января 2000 года?

Найдем общее количество более-менее современных книг (вышедших после 1-го января 2000-го года)

In [8]:
query1 = '''
SELECT
    COUNT (book_id) as cnt
FROM
    books
WHERE
    publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(query1, con = engine)

cnt
0  819

In [9]:
query11 = '''
SELECT
    MAX(publication_date) as max_date
FROM
    books
'''
pd.io.sql.read_sql(query11, con = engine)

max_date
0  2020-03-31

819 книг за 20 лет - не так уж и много - всего по 40 книг в год

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 1:</b> Тут все верно! 
</div>

### Количество обзоров и средняя оценка для каждой книги?

Теперь узнаем лидеров среди книг по числу обзоров и найдем для них среднюю оценку

In [19]:
query2 = '''
SELECT
    title,
    ROUND(AVG(rating), 2) as avg_rating,
    COUNT(DISTINCT review_id) as cnt_reviews
FROM
    books
    LEFT JOIN ratings ON books.book_id = ratings.book_id
    LEFT JOIN reviews ON books.book_id = reviews.book_id
GROUP BY
    books.book_id
ORDER BY
    cnt_reviews DESC
'''
pd.io.sql.read_sql(query2, con = engine).head(10)

title  avg_rating  cnt_reviews
0                            Twilight (Twilight  #1)        3.66            7
1                                Water for Elephants        3.98            6
2                                   The Glass Castle        4.21            6
3  Harry Potter and the Prisoner of Azkaban (Harr...        4.41            6
4  The Curious Incident of the Dog in the Night-Time        4.08            6
5             The Da Vinci Code (Robert Langdon  #2)        3.83            6
6                                      The Alchemist        3.79            6
7                The Hobbit  or There and Back Again        4.13            6
8                                     The Book Thief        4.26            6
9  The Lightning Thief (Percy Jackson and the Oly...        4.08            6

Как мы видим, самые популярные на отзывы Сумерки - книга с не самым высокой средней оценкой

<div class="alert alert-danger">
<b>Комментарий ревьюера v.1</b> 
    
📝 <b>Запрос 2:</b> Иван, этот запрос нужно немного скорректировать. Мы получаем слишком завышенные значения для столбца `cnt_reviews`. Это легко проверить: в таблице выше количество `cnt_reviews` = `26167`, а в исходной таблице `review` всего ревью `2793`.
    
Исправь, пожалуйста, запрос так, чтобы он не дублировал количество обзоров. 
    
Также, обрати, пожалуйста, внимание, что сейчас запрос выводит информацию о `999` книгах, но в базе данных их немного больше. А в запросе нас просят вывести информацию для всех книг, которые есть в базе. Возможно, нам нужно учитывать уникальный номер книги.      
</div> 

<div class="alert alert-info">
<b>Комментарий студента v.1</b>
    
Ошибочка вышла. Исправил

</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера v.2</b> 
    
👍 Запрос 2: Спасибо, теперь у нас нет дублирования записей. Но, 999 книг осталось потому, что в запросе мы используем title  для группировки данных. Лучше брать уникальные id книг, поскольку даже одно и тоже произведение могло издаваться разными издательствами или быть с разным переводом. Хороший пример - серии книг про Гарри Поттера, где есть несколько типов перевода и они оцениваются по разному.
</div>

<div class="alert alert-info">
<b>Комментарий студента v.2</b>
    
Вроде, теперь так, как должно быть

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.3</b> 
    
👍 <b>Запрос 1:</b> Теперь все верно!
</div>

In [11]:
# ячейка ревьюера:
query2 = '''
SELECT
    title,
    ROUND(AVG(rating), 2) as avg_rating,
    COUNT(DISTINCT review_id) as cnt_reviews
FROM
    books
    LEFT JOIN ratings ON books.book_id = ratings.book_id
    LEFT JOIN reviews ON books.book_id = reviews.book_id
GROUP BY
    books.book_id, title
ORDER BY
    cnt_reviews DESC
'''
pd.io.sql.read_sql(query2, con = engine).head(10)

title  avg_rating  cnt_reviews
0                            Twilight (Twilight  #1)        3.66            7
1                                Water for Elephants        3.98            6
2                                   The Glass Castle        4.21            6
3  Harry Potter and the Prisoner of Azkaban (Harr...        4.41            6
4  The Curious Incident of the Dog in the Night-Time        4.08            6
5             The Da Vinci Code (Robert Langdon  #2)        3.83            6
6                                      The Alchemist        3.79            6
7                The Hobbit  or There and Back Again        4.13            6
8                                     The Book Thief        4.26            6
9  The Lightning Thief (Percy Jackson and the Oly...        4.08            6

### Издательство, которое выпустило наибольшее число книг толще 50 страниц?

In [12]:
query3 = '''
SELECT
    publisher,
    COUNT(book_id) as cnt_books
FROM
    books
    LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE
    num_pages > 50
GROUP BY
    publisher
ORDER BY
    cnt_books DESC
'''
pd.io.sql.read_sql(query3, con = engine).head(10)

publisher  cnt_books
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25
3          Penguin Classics         24
4          Ballantine Books         19
5                    Bantam         19
6                   Berkley         17
7        St. Martin's Press         14
8             Berkley Books         14
9                     Delta         13

Перед нами десятка самых печатающих издательств. Пингвины впереди всех!

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 3:</b> Тут все корректно!
</div>

### Автор с самой высокой средней оценкой книг?

(книги только с 50 и более оценками)

In [13]:
query4 = '''
SELECT
    authors.author AS author,
    ROUND(AVG(rating), 2) AS avg_rating
FROM
    (SELECT
        books.book_id,
        author_id
    FROM
        books
        LEFT JOIN ratings ON books.book_id = ratings.book_id
    GROUP BY
        books.book_id
    HAVING
        COUNT(rating) >= 50) AS books_with_50_plus_ratings
    LEFT JOIN authors ON books_with_50_plus_ratings.author_id = authors.author_id
    LEFT JOIN ratings ON books_with_50_plus_ratings.book_id = ratings.book_id
GROUP BY
    author
ORDER BY
    avg_rating DESC
'''
pd.io.sql.read_sql(query4, con = engine).head(10)

author  avg_rating
0                         J.K. Rowling/Mary GrandPré        4.29
1                  Markus Zusak/Cao Xuân Việt Khương        4.26
2                                     J.R.R. Tolkien        4.25
3                                  Louisa May Alcott        4.19
4                                       Rick Riordan        4.08
5                                    William Golding        3.90
6                                      J.D. Salinger        3.83
7  William Shakespeare/Paul Werstine/Barbara A. M...        3.79
8           Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.79
9                                         Lois Lowry        3.75

Перед нами самы заметные авторы последнего времени. По ощущениям реальная картина схожая

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 4:</b> В этом запросе тоже все корректно. Обрати, пожалуйста, внимание, что в начале ты выделил книги, у которых 50 или больше оценок (тут все корректно), а затем подсчитал средний рейтинг для этих книг для каждого автора. Т.е. мы использовали все данные о рейтинге для всех книг. Для решения этого запроса мы могли подойти немного по-другому - в начале подсчитать рейтинг для книг, которые имеют 50 и более запросов, а потом уже подсчитать среднее значение по среднему рейтингу книг для каждого. Поэтому, советую учитывать при решении подобных задач такую вариативность. Последнее может быть актуально при большом количестве данных.
</div>

<div class="alert alert-info">
<b>Комментарий студента v.1</b>
    
Не совсем понял этот комментрий. Что значит "все данные о рейтинге"? И не понял, почему второй вариант эффективнее и как его реализовать

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 4:</b> Спасибо за вопрос и извини, если немного запутал. Сейчас, внутренним запросом мы определили книги, у которых есть 50 и более оценок. Тут все верно, далее к ним мы подключили данные по автору и рейтингу и в итоге получили массив данных, для всех книг у нас есть данные по рейтингу, а также по автору.
</div>

In [14]:
# ячейка ревьюера:
query4 = '''
SELECT *
FROM
  (SELECT books.book_id,
          author_id
   FROM books
   LEFT JOIN ratings ON books.book_id = ratings.book_id
   GROUP BY books.book_id
   HAVING COUNT(rating) >= 50) AS books_with_50_plus_ratings
LEFT JOIN authors ON books_with_50_plus_ratings.author_id = authors.author_id
LEFT JOIN ratings ON books_with_50_plus_ratings.book_id = ratings.book_id
'''
pd.io.sql.read_sql(query4, con = engine)

book_id  author_id  author_id           author  rating_id  book_id  \
0          75        106        106        Dan Brown        389       75   
1          75        106        106        Dan Brown        390       75   
2          75        106        106        Dan Brown        391       75   
3          75        106        106        Dan Brown        392       75   
4          75        106        106        Dan Brown        393       75   
...       ...        ...        ...              ...        ...      ...   
1408      948        554        554  Stephenie Meyer       6191      948   
1409      948        554        554  Stephenie Meyer       6192      948   
1410      948        554        554  Stephenie Meyer       6193      948   
1411      948        554        554  Stephenie Meyer       6194      948   
1412      948        554        554  Stephenie Meyer       6195      948   

           username  rating  
0        annawarner       5  
1          thomas49       4  
2           julie81       4  
3          robert00       4  
4        jarvispaul       4  
...             ...     ...  
1408  perrypatricia       3  
1409     carriehale       3  
1410       yvasquez       5  
1411  danielpatrick       4  
1412    carlsonbrad       5  

[1413 rows x 8 columns]

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 4:</b> Далее мы считаем среднее значение столбца `rating` для каждого автора. Давай немного рассмотрим возможные варианты. Может быть так, что у автора две книги, одна имеет 10 значений рейтинга - все пятерки (отличная книга), а вторая имеет всего 2 рейтинга - все двойки (неудачная книга, но и оценок меньше). В итоге у нас средний рейтинг (как мы это определили выше в запросе) будет (10*5 + 2*2) / 12 = 4.5. Но, мы можем подсчитать вначале средний рейтинг для каждой книги, а затем взять среднее значение. В нашем случае это будет как (10*5/10 + 2*2/2) / 2 = и среднее значение будет 3.5, что на балл ниже, чем при первом расчете. Давай вернемся к нашим книгам, как мы можем это реализовать:
</div>

In [15]:
# ячейка ревьюера: добавив дополнительную итерацию агрегации данных по книге:

query4 = '''

--- я взял твой запрос и сделал его временным запросом и группировал данные по книге, 
--- чтобы подсчитать средний рейтинг для каждой книги:
WITH book_rating AS
  (SELECT books_with_50_plus_ratings.book_id,
          authors.author AS author,
          AVG(rating) AS avg_rating
   FROM
     (SELECT books.book_id,
             author_id
      FROM books
      LEFT JOIN ratings ON books.book_id = ratings.book_id
      GROUP BY books.book_id
      HAVING COUNT(rating) >= 50) AS books_with_50_plus_ratings
   LEFT JOIN authors ON books_with_50_plus_ratings.author_id = authors.author_id
   LEFT JOIN ratings ON books_with_50_plus_ratings.book_id = ratings.book_id
   GROUP BY books_with_50_plus_ratings.book_id,
            authors.author) 
            
--- теперь обратимся к данным из временного запроса и подсчитаем средний рейтинг книг для каждого автора:
SELECT author AS author,
       AVG(avg_rating) AS avg_rating
FROM book_rating
GROUP BY author

'''
pd.io.sql.read_sql(query4, con = engine)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                                           Dan Brown    3.754540
2                                       J.D. Salinger    3.825581
3                                     William Golding    3.901408
4                                   Louisa May Alcott    4.192308
5            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
6                                     Stephenie Meyer    3.662500
7   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
8                                        Rick Riordan    4.080645
9             George Orwell/Boris Grabnar/Peter Škerl    3.729730
10                                     John Steinbeck    3.622951
11                                     J.R.R. Tolkien    4.258446
12                                         Lois Lowry    3.750000
13                  Markus Zusak/Cao Xuân Việt Khương    4.264151

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 4:</b> Как видишь, мы получили немного другие данные. Ниже мы можем это сравнить:
</div>

In [16]:
# ячейка ревьюера: добавив дополнительную итерации агрегации данных по книге:

query4 = '''

--- я взял твой запрос и сделал его временным запросом и группировал данные по книге, 
--- чтобы подсчитать средний рейтинг для каждой книги:
WITH book_rating AS
  (SELECT books_with_50_plus_ratings.book_id,
          authors.author AS author,
          AVG(rating) AS avg_rating
   FROM
     (SELECT books.book_id,
             author_id
      FROM books
      LEFT JOIN ratings ON books.book_id = ratings.book_id
      GROUP BY books.book_id
      HAVING COUNT(rating) >= 50) AS books_with_50_plus_ratings
   LEFT JOIN authors ON books_with_50_plus_ratings.author_id = authors.author_id
   LEFT JOIN ratings ON books_with_50_plus_ratings.book_id = ratings.book_id
   GROUP BY books_with_50_plus_ratings.book_id,
            authors.author),
     first_rating AS
  (SELECT authors.author AS author,
          AVG(rating) AS avg_rating
   FROM
     (SELECT books.book_id,
             author_id
      FROM books
      LEFT JOIN ratings ON books.book_id = ratings.book_id
      GROUP BY books.book_id
      HAVING COUNT(rating) >= 50) AS books_with_50_plus_ratings
   LEFT JOIN authors ON books_with_50_plus_ratings.author_id = authors.author_id
   LEFT JOIN ratings ON books_with_50_plus_ratings.book_id = ratings.book_id
   GROUP BY author) 
   
--- теперь обратимся к данным из временного запроса и подсчитаем средний рейтинг книг для каждого автора:
SELECT book_rating.author AS author,
       AVG(book_rating.avg_rating) AS avg_rating,
       COUNT(DISTINCT book_rating.book_id) AS book_count,
       AVG(first_rating.avg_rating) AS avg_rating_first
FROM book_rating
LEFT JOIN first_rating ON first_rating.author = book_rating.author
GROUP BY book_rating.author
ORDER BY avg_rating DESC

'''
pd.io.sql.read_sql(query4, con = engine)

author  avg_rating  book_count  \
0                          J.K. Rowling/Mary GrandPré    4.283844           4   
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151           1   
2                                      J.R.R. Tolkien    4.258446           2   
3                                   Louisa May Alcott    4.192308           1   
4                                        Rick Riordan    4.080645           1   
5                                     William Golding    3.901408           1   
6                                       J.D. Salinger    3.825581           1   
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474           1   
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879           1   
9                                           Dan Brown    3.754540           2   
10                                         Lois Lowry    3.750000           1   
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730           1   
12                                    Stephenie Meyer    3.662500           1   
13                                     John Steinbeck    3.622951           1   

    avg_rating_first  
0           4.287097  
1           4.264151  
2           4.246914  
3           4.192308  
4           4.080645  
5           3.901408  
6           3.825581  
7           3.789474  
8           3.787879  
9           3.741259  
10          3.750000  
11          3.729730  
12          3.662500  
13          3.622951

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 4:</b> Получается, там, где у нас количество книг будет больше одной, то средние рейтинги различаются. Если у нас было бы больше данных - различия могли бы быть более существенные. Я не говорю, что этот способ более эффективный, но будет не лишним посмотреть, какие альтернативные подходы могут быть использованы при решении этой задачи.
</div>

<div class="alert alert-info">
<b>Комментарий студента v.2</b>
    
Так понятнее, спасибо! Буду иметь в виду)
    
P.s. с оператором WITH ... AS ... просто ещё не работал (и в курсе его не было👀)

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.3</b> 
    
👍 <b>Запрос 4:</b> Здорово, что теперь стало яснее. По поводу WITH ... AS - тут логика такая - мы делаем временную таблицу, к которой потом обращаемся. Можно взять за аналог создание дополнительного среза из датасета pandas. Все тоже самое, можно было бы решить, используя вложенные запросы. Тут, наверно, кому как удобнее. 
</div>

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок?

In [17]:
query5 = '''
SELECT
    ROUND(AVG(cnts_reviews.cnt_reviews), 2) AS avg_cnt_reviews
FROM
    (SELECT
        COUNT(review_id) AS cnt_reviews
    FROM (
        SELECT
            username
        FROM
            ratings
        GROUP BY
            username
        HAVING
            COUNT(rating_id) > 50
            ) AS users_with_50_plus_ratings
        LEFT JOIN reviews ON users_with_50_plus_ratings.username = reviews.username
    GROUP BY
        reviews.username) AS cnts_reviews
'''
pd.io.sql.read_sql(query5, con = engine)

avg_cnt_reviews
0            24.33

Количество отзывов даже у самых активных пользователей почти в 2 раза меньше, чем количество оценок. Не пора ли простимулировать пользователей писать больше отзывов?

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 5:</b> Здесь тоже все корректно! Обрати, пожалуйста, внимание, что мы можем округлить до 2-3 знаков после запятой, для удобства.
</div>

<div class="alert alert-info">
<b>Комментарий студента v.1</b>
    
Округлил

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 5:</b> Спасибо!
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера v.1</b> 
    
📝 Несмотря на то, что это не требуется заданием - лучше в конце проекта, даже такого небольшого, как этот, оставить резюме по поводу проделанной работы. Так проект / отчет выглядит более завершенным, да и коллеженки часто спрашивают, когда им пересылаешь отчет - "расскажи в двух словах, чего там".
    
</div>

**Вывод:**
- за последние 20 лет каждый год появлялось, в среднем, по 40 книг;
- cамые популярные по числу отзывов книги: Мемуары гейши(8 отзывов и средняя оценка 4.14) и Сумерки(7 и 3.66);
- самые выпускающее издательство - Penguin (Penguin Books + Penguin Classics) - выпустили 66 книг;
- лучшие писатели по среднему рейтингу книг - Джоан Роалинг, Маркус Зусак и Джон Толкин;
- среди самых активных пользователей (тех, кто поставил больше 50 оценок) среднее количество написанных отзывов равно 24.33 - то есть менее, чем на каждую вторую оцененную книгу

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 5:</b> Вышло - отлично, спасибо!
</div>